### ISSUE TITLE
PNG driver alpha range issue for 16bit PNG

### ISSUE THREAD

**Issue by trinopoty** [OSGeo/gdal#2267]
Color channels in 16bit PNG files have a range of 0x0000 - 0xFFFF. But, GDAL clamps the alpha channel to the range 0x00-0xFF. This results in an almost transparent image from the perspective of any reading application.

This should be fixed for 16 bit PNG files to make the alpha channel use all 16 bits instead of just 8 bits.

**Comment by trinopoty** [comment]
I would be willing to submit a PR for this if someone were to point me where the changes should be done.

**Comment by jratike80** [comment]
Could you add some small sample image and list a few of the "any reading applications" for making testing easier?

**Comment by trinopoty** [comment]
You can see the files: [normal](https://imgur.com/a/uz0rRHL) and [fixed](https://imgur.com/a/H5tlFSd.
The original tiff was created with 2 (data + alpha) channels of type UInt16. The alpha values were trimmed to the range 0x00-0xFF by the GTiff driver and this was then copied over to the PNG making the PNG transparent.
The two files are the same files except I've modified the `fixed` file to set the alpha to 0xFFFF.
The effect should be visible to any programs including your web browser.

**Comment by trinopoty** [comment]
I have also tried code to the effect of:
`
alpha_data = dataset.GetRasterBand(2) # This is my alpha band
alpha_data = wr_alpha_band.GetVirtualMemArray(eAccess=gdal.GA_Update)
alpha_data = (alpha_data / 255) * 65535

out_drv = gdal.GetDriverByName('PNG')
out_drv.CreateCopy(filename, dataset, strict=0)
`

But the output PNG files still have the incorrect alpha values.

**Comment by trinopoty** [comment]
As far as I can tell, GDAL internally clamps alpha channel to 0x00 to 0xFF. I remember reading this somewhere in the documentation but I can't find it at the moment.

**Comment by jratike80** [comment]
Verified that IrfanView, default Windows Photo viewer, and GIMP do not show the "normal" file correctly.

QGIS seems to be an exception and it shows also the "normal" image more or less right. Small difference is visible at the boundary of nodata between "normal" and "fixed".

**Comment by jratike80** [comment]
This may be related https://github.com/OSGeo/gdal/pull/742 and this as well
https://trac.osgeo.org/gdal/wiki/rfc15_nodatabitmask

It could be that a more thorough code review is needed for checking how non-byte alpha band behaves within GDAL.

**Comment by trinopoty** [comment]
You can see [here](https://gdal.org/api/raster_c_api.html#_CPPv413GCI_AlphaBand) that values are clamped between 0x00 and 0xFF for bands of type `GCI_AlphaBand`.
At least, the C documentation seems to suggest so.

**Comment by trinopoty** [comment]
Also, I'm not using internal masks so I don't think this issue is related to #742 but I may be wrong. Particularly if GDAL optimizes the alpha band to a bitmask at some point.

**Comment by rouault** [comment]
I'm confused by this report
On the 'fixed' image https://imgur.com/a/H5tlFSd , we get:
```
$ gdalinfo -mm /tmp/fixed.png 
Driver: PNG/Portable Network Graphics
Files: /tmp/fixed.png
Size is 256, 256
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  256.0)
Upper Right (  256.0,    0.0)
Lower Right (  256.0,  256.0)
Center      (  128.0,  128.0)
Band 1 Block=256x1 Type=UInt16, ColorInterp=Gray
    Computed Min/Max=0.000,64651.000
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=256x1 Type=UInt16, ColorInterp=Alpha
    Computed Min/Max=0.000,65535.000
```
So the alpha values are 0 and 65535 as expected

And when doing `gdal_translate /tmp/fixed.png out.png`, then we get the same on the result:
```
$ gdalinfo -mm out.png
Driver: PNG/Portable Network Graphics
Files: out.png
Size is 256, 256
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  256.0)
Upper Right (  256.0,    0.0)
Lower Right (  256.0,  256.0)
Center      (  128.0,  128.0)
Band 1 Block=256x1 Type=UInt16, ColorInterp=Gray
    Computed Min/Max=0.000,64651.000
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=256x1 Type=UInt16, ColorInterp=Alpha
    Computed Min/Max=0.000,65535.000
```

**Comment by trinopoty** [comment]
I fixed the `fixed` image with a small C program leveraging libpng.
Also, I've experimented with this a little more and found that the PNG driver can indeed handle 16 bit alpha properly. The problem seems to be with the GTIFF driver that trims the alpha channel values.
But, I wouldn't be surprised if that's because the channel was converted to a bitmap internally and later expanded into 0 and 255 when reading.

Basically, doing this fixes the alpha issue:
```
alpha_data = wr_alpha_band.ReadAsArray()
alpha_data = (alpha_data / 255) * 65535
alpha_data = alpha_data.astype(numpy.uint16)
wr_alpha_band.WriteArray(alpha_data)
```

**Comment by jratike80** [comment]
Test on a GDT_UInt16 RGBA (#5692) in 
https://github.com/OSGeo/gdal/blob/master/autotest/gcore/mask.py
is testing also "Test special case where dynamics is only 0-255".

GDAL seems to consider sometimes alpha channel just as a binary mask and then it only checks if the value of alpha is 0 or >0.

**Comment by rouault** [comment]
Can't reproduce this with the GTiff driver either:
```
even@even-i700:~/gdal/git/gdal/gdal$ gdal_translate out.png out.tif
Input file size is 256, 256
0...10...20...30...40...50...60...70...80...90...100 - done.

even@even-i700:~/gdal/git/gdal/gdal$ gdalinfo -mm out.tif
Driver: GTiff/GeoTIFF
Files: out.tif
Size is 256, 256
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  256.0)
Upper Right (  256.0,    0.0)
Lower Right (  256.0,  256.0)
Center      (  128.0,  128.0)
Band 1 Block=256x8 Type=UInt16, ColorInterp=Gray
    Computed Min/Max=0.000,64651.000
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=256x8 Type=UInt16, ColorInterp=Alpha
    Computed Min/Max=0.000,65535.000
```

Please be very precise in what you do. You also didn't mention which GDAL version you use. My tests are with master (but likely to give identical reasons with 2.4)

**Comment by trinopoty** [comment]
Could be a problem somewhere in my own code. But, I've found a viable workaround for the issue I had so I'm closing this issue.

**Comment by jratike80** [comment]
If you'll find out what really happened please add a comment. It could help other users in the future.

**Comment by trinopoty** [comment]
I was trying to convert a tiff with a single data band into a tiff with a uint16 data band and alpha band with some additional processing.
My code is like this:
```
    for y in range(0, input_band.YSize):
        data = input_band.GetVirtualMemArray(
            eAccess=gdal.GF_Read, 
            xoff=0,
            yoff=y,
            xsize=input_band.XSize,
            ysize=1,
            bufxsize=input_band.XSize,
            bufysize=1,
            datatype=input_band.DataType)
        
        # Extend the min,max into 0,65535
        output_array = np.zeros((1, input_band.XSize), dtype=np.uint16)
        output_array = output_array + ((((data - val_min) / val_den) * 65535) * (data != nodata))

        mask_array = np.zeros((1, input_band.XSize), dtype=np.uint16)
        mask_array = mask_array + (65535 * (data != nodata))

        output_band.WriteArray(output_array, xoff=0, yoff=y)
        output_mask_band.WriteArray(mask_array, xoff=0, yoff=y)
```

Even though I set the values in the alpha channel to be 0 or 65535, the output tiff only had alpha values 0 and 255 when I opened and read it again in a different script.
